llama/model.py

```diff
@@ -190,99 +212,167 @@ def __init__(
+            world_size = get_model_parallel_world_size()
+            rank = get_model_parallel_rank()
+
+        init_method = lambda x: x
+
         self.w1 = ColumnParallelLinear(
-            dim, hidden_dim, bias=False, gather_output=False, init_method=lambda x: x
+            dim, hidden_dim, bias=False, gather_output=False, init_method=init_method,
+            world_size=world_size, rank=rank, groups=groups
         )
         self.w2 = RowParallelLinear(
-            hidden_dim, dim, bias=False, input_is_parallel=True, init_method=lambda x: x
+            hidden_dim, dim, bias=False, input_is_parallel=True, init_method=init_method,
+            world_size=world_size, rank=rank, groups=groups
         )
         self.w3 = ColumnParallelLinear(
-            dim, hidden_dim, bias=False, gather_output=False, init_method=lambda x: x
+            dim, hidden_dim, bias=False, gather_output=False, init_method=init_method,
+            world_size=world_size, rank=rank, groups=groups
         )
 
+        # self.w1 = ColumnParallelLinear(
+        #     dim, hidden_dim, bias=False, gather_output=False, init_method=lambda x: x
+        # )
+        # self.w2 = RowParallelLinear(
+        #     hidden_dim, dim, bias=False, input_is_parallel=True, init_method=lambda x: x
+        # )
+        # self.w3 = ColumnParallelLinear(
+        #     dim, hidden_dim, bias=False, gather_output=False, init_method=lambda x: x
+        # )
+
     def forward(self, x):
         return self.w2(F.silu(self.w1(x)) * self.w3(x))
 
 
 class TransformerBlock(nn.Module):
-    def __init__(self, layer_id: int, args: ModelArgs):
+    def __init__(self, layer_id: int, args: ModelArgs, world_size: Optional[int] = None,
+                 rank: Optional[int] = None, groups: Optional[List] = None):
         super().__init__()
         self.n_heads = args.n_heads
         self.dim = args.dim
         self.head_dim = args.dim // args.n_heads
-        self.attention = Attention(args)
+
+        if world_size is None:
+            groups = get_model_parallel_group()
+            world_size = get_model_parallel_world_size()
+            rank = get_model_parallel_rank()
+
+        self.attention = Attention(args, world_size=world_size,
+                                   rank=rank, groups=groups)
         self.feed_forward = FeedForward(
             dim=args.dim,
             hidden_dim=4 * args.dim,
             multiple_of=args.multiple_of,
-            ffn_dim_multiplier=args.ffn_dim_multiplier,
+            world_size=world_size, rank=rank, groups=groups,
+            ffn_dim_multiplier=args.ffn_dim_multiplier
         )
-        self.layer_id = layer_id
-        self.attention_norm = RMSNorm(args.dim, eps=args.norm_eps)
-        self.ffn_norm = RMSNorm(args.dim, eps=args.norm_eps)
+        self.layer_id = layer_id #NEW LINE
+        self.attention_norm = RMSNorm(args.dim, eps=args.norm_eps) #NEW LINE
+        self.ffn_norm = RMSNorm(args.dim, eps=args.norm_eps) #NEW LINE
```

Comment:
What does `#NEW LINE` mean? They seem to be same in LLaMA 1.

Reply:
lol - don't recall - they indeed exist in the previous generation